packages/grafana-data/src/transformations/transformers/labelsToFields.ts

```diff
@@ -1,173 +1,61 @@
 import { DataFrame, DataTransformerInfo, FieldType, Field } from '../../types';
 import { DataTransformerID } from './ids';
 import { ArrayVector } from '../../vector';
-import { filterFieldsTransformer } from './filter';
-import { FieldMatcherID } from '..';
-import { MutableField } from '../../dataframe';
+import { mergeTransformer } from './merge';
 
-export interface LabelsToFieldsOptions {}
-type MapItem = { type: FieldType; values: Record<string, any>; isValue: boolean };
-type SeriesMapItem = Record<string, MapItem>;
-type Map = Record<string, SeriesMapItem>;
+export interface LabelsToFieldsOptions {
+  /*
+   * If set this will use this label's value as the value field name.
+   */
+  valueLabel?: string;
+}
 
 export const labelsToFieldsTransformer: DataTransformerInfo<LabelsToFieldsOptions> = {
   id: DataTransformerID.labelsToFields,
   name: 'Labels to fields',
-  description: 'Groups series by time and return labels as columns',
+  description: 'Extract time series labels to fields (columns)',
   defaultOptions: {},
   transformer: options => (data: DataFrame[]) => {
-    const framesWithTimeField = filterFieldsTransformer.transformer({ include: { id: FieldMatcherID.time } })(data);
-    if (!framesWithTimeField.length || !framesWithTimeField[0].fields.length) {
-      return data;
-    }
-
-    // get frames with only value fields
-    const framesWithoutTimeField = getFramesWithOnlyValueFields(data);
-    if (!framesWithoutTimeField.length || !framesWithoutTimeField[0].fields.length) {
-      return data;
-    }
+    const result: DataFrame[] = [];
 
-    const columnsMap = createColumnsMap(framesWithTimeField, framesWithoutTimeField);
-    const fields = createFields(columnsMap);
-    const values: Record<string, any[]> = {};
+    for (const frame of data) {
+      const newFields: Field[] = [];
 
-    const timeColumnItem = columnsMap[fields[0].name];
-    const seriesIndexStrings = Object.keys(timeColumnItem);
-    for (const seriesIndexString of seriesIndexStrings) {
-      const seriesItem = timeColumnItem[seriesIndexString];
-      const timeValueStrings = Object.keys(seriesItem.values);
+      for (const field of frame.fields) {
+        if (field.labels) {
```

Comment:
nitpick: maybe reverse `if` statement and continue to minimize nesting?

Reply:
fixed